tests/integration/test_multiple_data.py

```diff
@@ -104,25 +105,33 @@ def test_multiple_dataframes_eight_schools_error_to_many_to_few_rows():
     data_df = pandas.read_csv(os.path.join(test_dir, "eight_schools.csv"))
 
     y_data_df = data_df[["y", "school"]]
-    sigma_data_df = pandas.concat([data_df[["school", "sigma"]], data_df[["school", "sigma"]]])
+    sigma_data_df = pandas.concat([data_df[["school", "sigma"]], data_df[["school", "sigma"]].assign(sigma=5.0)])
 
     model_string = """
-    y' ~ normal(theta[school], sigma[school]);
-    theta' = mu + z[school] * tau;
-    z ~ normal(0, 1);
+    y[school]' ~ normal(theta[school], sigma[school]);
+    theta[school]' = mu + z[school] * tau;
+    z[school] ~ normal(0, 1);
     mu ~ normal(0, 5);
     tau<lower = 0.0> ~ log_normal(0, 1);
     """
 
-    with pytest.raises(Exception, match="unique"):
-        Model({"y_data": y_data_df, "sigma_data": sigma_data_df}, model_string=model_string)
+    # https://stackoverflow.com/questions/70499076/how-to-check-exception-cause-using-pytest-raises
```

Comment:
Since we're using `raise E from e` alot to pass around exceptions, pytests can't match exception messages from the inner cause exceptions.

In addition, I've found there's too many intermediate phases in the stack trace, which sorta "hides" the original(cause) exception in a wall of text, potentially making it hard to find.

Reply:
Makes sense. Looking at the code there, we could:
1. Throw the error in the bind function
2. Throw the error from code calling bind

Throwing the error in the bind gives us the most specific error information.

Throwing from the code calling the bind gives us access to the ast (bind doesn't know about this).

I believe we want both things. Without passing the ast through to bind somehow, what if we adopted an add_note strategy: https://peps.python.org/pep-0678/

The lingo then would be instead of re-raising unrecoverable errors that we just append information as best we can and let the original exception propagate. In this case we'd want a way to do the AST print without using CompileError.

Reply:
Interesting. What I think is necessary is printing the source code(where it messed up) and the actual informal error message as one.

Notes honestly sounds like a glorified 'append text to exception message' to me, but I think it will serve our purpose here(and if it didn't exist I would've proposed to just append the error meswage to the new exception message)